### ISSUE TITLE
Thousands of SVG cache files

### ISSUE THREAD

**Issue by SammysHP** [cgeo/cgeo#12145]
### Describe your problem!

I've got 63,702 PNG files named `svg-\d+.png` in `/data/data/cgeo.geocaching/files`. They contain icons. Dates range from the present back to beginning of the year.

### How to reproduce?

…

### Reproducible

Unclear

### c:geo Version

Doesn't matter, various versions

**Comment by SammysHP** [comment]
Looks like they come from mapsforge as they are created when I open the map.

The files are usually very small, but take lots of disk space due to the minimum allocation size of 4 KB.

**Comment by Lineflyer** [comment]
As my device is not rooted I cannot look into that directory, however on the emulator I can't see such files while using mapsforge:
![image](https://user-images.githubusercontent.com/949669/142757575-08025435-9924-495a-a0da-db952d6e15e9.png)

**Comment by eddiemuc** [comment]
I can confirm this on my emulator unfortunately. It is a Pixel 3a emulator using SDK29 (Android 10.0 / Q). Standard emulator where I took over all default settings.

![image](https://user-images.githubusercontent.com/6909759/142758158-89f7b1d6-33df-4b40-a017-a8bd75b56af9.png)

**Comment by eddiemuc** [comment]
I checked this in mapsforge code. There is a caching mechanism for usage of svgs installed which apparently can't be disabled. However, we can configure the place where images are cached by using method `AndroidGraphicFactory.setSvgCacheDir`.

To avoid duplicate storage, mapsforge uses a hasing algorithm to calculate the id you see in the filename. This hash algorithm includes filename (as used in theme file), width, height, percentage for scaling) and resourceProvider. Thus the list of svgs is expected to grow if you update theme files (and providers change the image names in the underlying xmls) or the method how themes are accessed (e.g. zipped theme files vs direct access). However, the list should not grow "endlessly".

The actual caching logic is implemented here:
https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map-android/src/main/java/org/mapsforge/map/android/graphics/AndroidSvgBitmapStore.java 

The setting of the app-specific svg cache dir can be done here:
https://github.com/mapsforge/mapsforge/blob/8d1e278eadb065b811cc8c5b8000f13394d33f42/mapsforge-map-android/src/main/java/org/mapsforge/map/android/graphics/AndroidGraphicFactory.java#L318-L320

The hasing algorithm iss implemented here:

https://github.com/mapsforge/mapsforge/blob/8d1e278eadb065b811cc8c5b8000f13394d33f42/mapsforge-map/src/main/java/org/mapsforge/map/rendertheme/XmlUtils.java#L60-L66

**Comment by eddiemuc** [comment]
Update: I think I found the problem. For resourceProvider the hashcode is used, but resource provider does not implement this. So I guess hash value for this changes with every app start. Must be fixed in mapsforge code, I will create a PR for it.

**Comment by eddiemuc** [comment]
I need some git/github help on this one.

I created a commit on my own fork of mapsforge at eddiemuc/mapsforge in a new branch "issues/svg-cache-not-working-with-resource-providers"
* Branch: https://github.com/eddiemuc/mapsforge/tree/issues/svg-cache-not-working-with-resource-providers 
* Commit: https://github.com/eddiemuc/mapsforge/commit/77a4b35228067f342ddd47efe7b4519db583e629 

I now want to get this commit onto the cgeo fork of mapsforge at cgeo/mapsforge, branch "fix-potential-indexoutofbound-on-parallel-processing "
* Branch https://github.com/cgeo/mapsforge/tree/fix-potential-indexoutofbound-on-parallel-processing 

I tried to open a pull request against cgeo/mapsforge branch "fix-potential-indexoutofbound-on-parallel-processing" but apparently this seems not to be possible.

How can I do this then instead?

**Comment by eddiemuc** [comment]
@Lineflyer I tried pushing to cgeo/mapsforge directly, but lack the necessary permissions

**Comment by moving-bits** [comment]
Isn't the fix for "out of bounds" already merged to our fork of mapsforge? Then you would be able to open a PR against the main branch of our mapsforge fork, no need to PR against a sub branch (which may not be enabled for opening PR).

**Comment by eddiemuc** [comment]
> Isn't the fix for "out of bounds" already merged to our fork of mapsforge? Then you would be able to open a PR against the main branch of our mapsforge fork, no need to PR against a sub branch (which may not be enabled for opening PR).

Good point, but I am unable to open a PR against cgeo/mapsforge "master" either. However I try to open a Pull Request, "cgeo/mapsforge" is not offeres as choice for base repo:

![image](https://user-images.githubusercontent.com/6909759/142761199-cb268112-eb85-4f65-abff-3f5179a837dc.png)

![image](https://user-images.githubusercontent.com/6909759/142761211-c517317e-a437-4605-861e-040dcdf6bad6.png)

![image](https://user-images.githubusercontent.com/6909759/142761225-ca087a00-fead-4ae4-816a-ed52f1181165.png)

**Comment by eddiemuc** [comment]
I assume I have no permission to issue PRs against cgeo/mapsforge...

**Comment by moving-bits** [comment]
Maybe @Lineflyer needs to enable creating PR against base `cgeo/mapsforge` first?
(Although then I don't understand how we've been able to include the other fix... 🤔)

**Comment by Lineflyer** [comment]
Strange... there are AFAIK no permission settings for PRs to public repositories. Will doublecheck anyway.

**Comment by moving-bits** [comment]
Or maybe you need to fork `cgeo/mapsforge` and build your PR against that fork to be able to PR against `cgeo/mapsforge`?

**Comment by Lineflyer** [comment]
I set up same permissions to cgeo/mapsforge as we have for cgeo/cgeo.

@eddiemuc 
I would suggest to add cgeo/mapsforge as a remote (via `git remote add ....`) and just push your fix branch and continue from there.

**Comment by eddiemuc** [comment]
> @eddiemuc I would suggest to add cgeo/mapsforge as a remote (via `git remote add ....`) and just push your fix branch and continue from there.

Tried that but failed. I have no permission to push to cgeo/mapsforge.

**Comment by Lineflyer** [comment]
> Update: seems to work now, push was successful

Assigned permissions just some minutes ago.
I assumed it would take the basic role to allow cgeo/team members, but it seems it need to be setup explicitly.

**Comment by eddiemuc** [comment]
It worked. Have modified + pushed PR accordingly.

**Comment by Lineflyer** [comment]
Just for the records:
Found the fault: The base role was set to "No permission" in https://github.com/organizations/cgeo/settings/member_privileges
Therefore no one was able to interact (However: Still wondering why not even submitting a PR was possible).

Now I have set writing permission to cgeo/team and cgeo/core (whereas pushing to `master` is restricted to cgeo/core), admin permission to cgeo/admins and enabled branch protection for `master` (no force pushes). (=Same settings as in cgeo/cgeo).
